File: problems-originals/3801-3900/3887_incremental-even-weighted-cycle-queries/solution_color_flipping.py
class Solution:
    def numberOfEdgesAdded(self, n: int, edges: list[list[int]]) -> int:
        parent = list(range(n))
        size = [1] * n
        color = [0] * n  # absolute color of each node within its component
        members = [[i] for i in range(n)]  # per-root member lists, for the flip

        def find(x):  # membership only: path halving, no parity bookkeeping
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x

        added = 0
        for u, v, w in edges:
            ru, rv = find(u), find(v)
            if ru == rv:
                # the standing path parity is color[u] ^ color[v]: an O(1) verdict
                if color[u] ^ color[v] == w:
                    added += 1
            else:
                if size[ru] < size[rv]:
                    ru, rv = rv, ru  # ru is now the larger root
                if color[u] ^ color[v] != w:
                    # recolor the smaller component: every relation inside it
                    # survives a uniform flip, while the new edge's demand flips
                    for m in members[rv]:
                        color[m] ^= 1
                parent[rv] = ru
                size[ru] += size[rv]
                members[ru].extend(members[rv])
                added += 1
        return added

File: problems-adapt/3801-3900/3887_keep-every-cycle-even/solution_color_flipping.rs
impl Solution {
    pub fn edges_admitted(n: i32, edges: Vec<Vec<i32>>) -> i32 {
        let n = n as usize;
        let mut parent: Vec<usize> = (0..n).collect();
        let mut size = vec![1usize; n];
        let mut color = vec![0i32; n]; // absolute color of each node within its component
        let mut members: Vec<Vec<usize>> = (0..n).map(|i| vec![i]).collect(); // per-root member lists

        // membership only: path halving, no parity bookkeeping
        let find = |parent: &mut Vec<usize>, x: usize| -> usize {
            let mut cur = x;
            while parent[cur] != cur {
                parent[cur] = parent[parent[cur]];
                cur = parent[cur];
            }
            cur
        };

        let mut added = 0i32;
        for e in &edges {
            let u = e[0] as usize;
            let v = e[1] as usize;
            let w = e[2];
            let mut ru = find(&mut parent, u);
            let mut rv = find(&mut parent, v);
            if ru == rv {
                // the standing path parity is color[u] ^ color[v]: an O(1) verdict
                if (color[u] ^ color[v]) == w {
                    added += 1;
                }
            } else {
                if size[ru] < size[rv] {
                    std::mem::swap(&mut ru, &mut rv); // ru is now the larger root
                }
                if (color[u] ^ color[v]) != w {
                    // recolor the smaller component: every relation inside it
                    // survives a uniform flip, while the new edge's demand flips
                    for &m in members[rv].iter() {
                        color[m] ^= 1;
                    }
                }
                parent[rv] = ru;
                size[ru] += size[rv];
                let absorbed = std::mem::take(&mut members[rv]);
                members[ru].extend(absorbed);
                added += 1;
            }
        }
        added
    }
}
